### PR TITLE
DEP: scipy.stats.morestats: clarify deprecation warnings

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-17572
Toward gh-17771
Toward gh-18279

#### What does this implement/fix?
1. gh-17572 and gh-17771 report inaccurate deprecation warnings. An equivalent example from `scipy.stats.morestats` is:

```python3
from scipy.stats.morestats import wilcoxon  # message is appropriate
# DeprecationWarning: Please use `wilcoxon` from the `scipy.stats` namespace, 
# the `scipy.stats.morestats` namespace is deprecated.

from scipy.stats import wilcoxon  # import works as expected

from scipy.stats.morestats import WilcoxonResult  # message is inaccurate
# DeprecationWarning: Please use `WilcoxonResult` from the `scipy.stats` namespace, 
# the `scipy.stats.morestats` namespace is deprecated.

from scipy.stats import WilcoxonResult  # this fails, as it should
# ImportError: cannot import name 'WilcoxonResult' from 'scipy.stats'
```

The problem is that the deprecation warning logic does not distinguish between:
- public attributes that are now to be imported only from the subpackage
  (e.g. `scipy.stats.morestats.wilcoxon` should only be imported as `scipy.stats.wilcoxon`)
- private attributes that are not supposed to be imported at all
  (e.g. `scipy.stats.morestats.WilcoxonResult` was never  public, and should no longer be imported at all)

A solution is to distinguish between the two. With this PR:
```python3
from scipy.stats.morestats import WilcoxonResult

# DeprecationWarning: `scipy.stats.morestats.WilcoxonResult` is deprecated along with the 
# `scipy.stats.morestats` namespace. 
# `scipy.stats.morestats.WilcoxonResult` will be removed in SciPy 1.13.0, and the
# `scipy.stats.morestats` namespace will be removed in SciPy 2.0.0.
```

With this timeline established, we can remove private attributes like `WilcoxonResult` from the `__all__` of these modules in 1.13.0.

The logic for public attributes is unchanged, but the deprecation message is more specific about when the module will be removed.

```python3
from scipy.stats.morestats import wilcoxon
# DeprecationWarning: Please import `wilcoxon` from the `scipy.stats` namespace; 
# the `scipy.stats.morestats` namespace is deprecated and will be removed in SciPy 2.0.0.
```

It sounds like there was a decision to remove these namespaces in SciPy 2.0.0, so `import scipy.stats.morestats` will work until 2.0.0, but was there a decision that that `from scipy.stats.morestats import wilcoxon` has to work until then? We could adjust the message to state that `scipy.stats.morestats.wilcoxon` will be removed in 1.13.0 and the whole namespace will be removed in 2.0.0. 

---

2. When `getattr`ing something from one of these modules that does not exist, we get an erroneous error message.

```python3
import scipy.stats.morestats
getattr(scipy.stats.morestats, 'ekki')
# AttributeError: scipy.stats.morestats is deprecated and has no attribute ekki. 
# Try looking in scipy.stats instead.
```

Strings that result in this message are *never* present in the subpackage namespace, as far as I can tell, so there is no reason to suggest looking there. Now:

```python3
import scipy.stats.morestats
getattr(scipy.stats.morestats, 'ekki')
# AttributeError: `scipy.stats.morestats` has no attribute `ekki`; furthermore,
# `scipy.stats.morestats` is deprecated and will be removed in SciPy 2.0.0.
```

#### Additional information
The logic here is easily abstracted to work for any module/subpackage by specifying the appropriate strings. The implementation could live in `_lib` so we can avoid copy-paste and have a consistent message format throughout SciPy. That is work for a follow-up.

### CODE REVIEW

**Comment by mdhaber** [comment]
Test passed locally before, then CI merged main and they fail. Looks like this is because `atleast_1d` is no longer imported in `scipy.stats._morestats`. Without a test like the one introduced in this PR, attributes may be removed accidentally during the deprecation period.

**Comment by mdhaber** [comment]
@h-vetinari @j-bowhay thought this might be up your alley. If not of interest, no problem; I'll go ahead and close.

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -32,11 +33,22 @@ def __dir__():
...
+    attr = getattr(scipy.stats, name, None)
+
+    if attr is not None:
+        message = (f"Please import `{name}` from the `scipy.stats` namespace; "
+                   "the `scipy.stats.morestats` namespace is deprecated and "
+                   "will be removed in SciPy 2.0.0.")
+    else:
+        message = (f"`scipy.stats.morestats.{name}` is deprecated along with "
+                   "the `scipy.stats.morestats` namespace. "
+                   f"`scipy.stats.morestats.{name}` will be removed in SciPy 1.13.0, and "
```
Comment (j-bowhay) [inline scipy/stats/morestats.py:49]:
```suggestion
                   f"`scipy.stats.morestats.{name}` will be removed in SciPy 1.14.0, and "
```

**Review by j-bowhay — COMMENTED** [review]
Thanks @mdhaber will give this proper look soon

File: scipy/stats/morestats.py (hunk 1 above)
Comment (h-vetinari) [reply scipy/stats/morestats.py:49]: So, I think this reaction is to our usual "2 versions from deprecation warning to removal" policy, but even though it's not obvious from the diff in this case, all(?) of these are raising deprecation warnings already, and we're only sharpening it. This is in line with how we've done this for other long-existing deprecations, and I'm fine to announce removal for 1.13.

```
>>> import scipy
>>> scipy.__version__
'1.11.1'
>>> from scipy.stats.morestats import wilcoxon
<stdin>:1: DeprecationWarning: Please use `wilcoxon` from the `scipy.stats` namespace, the `scipy.stats.morestats` namespace is deprecated.
>>> from scipy.stats import wilcoxon
>>> from scipy.stats.morestats import WilcoxonResult
<stdin>:1: DeprecationWarning: Please use `WilcoxonResult` from the `scipy.stats` namespace, the `scipy.stats.morestats` namespace is deprecated.
>>> from scipy.stats import WilcoxonResult
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'WilcoxonResult' from 'scipy.stats' (C:\Users\axelo\.conda\envs\dev\lib\site-packages\scipy\stats\__init__.py)
```

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -32,11 +33,22 @@ def __dir__():
 def __getattr__(name):
     if name not in __all__:
         raise AttributeError(
-            "scipy.stats.morestats is deprecated and has no attribute "
-            f"{name}. Try looking in scipy.stats instead.")
-
-    warnings.warn(f"Please use `{name}` from the `scipy.stats` namespace, "
-                  "the `scipy.stats.morestats` namespace is deprecated.",
-                  category=DeprecationWarning, stacklevel=2)
```
Comment (h-vetinari) [inline scipy/stats/morestats.py:40]: In particular, this warning _always_ fired once we're in this function.

**Review by h-vetinari — APPROVED** [review]
Thanks for the ping @mdhaber!

The plan sounds sensible to me. I haven't gone over it with a fine-toothed comb, but it looks good to me.

File: scipy/stats/morestats.py (hunk 1 above)
Comment (j-bowhay) [reply scipy/stats/morestats.py:49]: Good point

**Comment by h-vetinari** [comment]
> Without a test like the one introduced in this PR, attributes may be removed accidentally during the deprecation period. Shall I add `atleast_1d` back to `_morestats.py` to make the test pass, or should it be removed from `__all__` of `morestats.py`?

Sorry, didn't notice this aspect before. I'd say add it back for now, but no strong preference.

**Comment by mdhaber** [comment]
Thanks @j-bowhay @h-vetinari. I added `atleast_1d` back to `_morestats.py` to fix the failure, and the remaining failure is unrelated. (I'll fix that in another PR shortly.)

**Review by j-bowhay — APPROVED** [review]
Thanks @mdhaber this looks like a good improvement to me. Happy to add the follow-up to my to-do list unless you already have it in the works

**Comment by mdhaber** [comment]
No, it would be great if you'd like to follow up with the generic version.

**Pull request merged**